pytorch_lightning/core/optimizer.py

```diff
@@ -60,17 +60,40 @@ def __init__(self,
         self._trainer = None
         self._optimizer = optimizer
         self._accumulate_grad_batches = accumulate_grad_batches
-        self._use_accumulate_grad_batches_from_trainer = accumulate_grad_batches is None
+        self._automatic_optimization = None
+        self._optimizer_idx = None
+
+    @property
+    def accumulate_grad_batches(self):
+        return self._accumulate_grad_batches
+
+    @accumulate_grad_batches.setter
+    def accumulate_grad_batches(self, accumulate_grad_batches):
+        if self._automatic_optimization is None or self._automatic_optimization:
+            raise MisconfigurationException(
+                'In automatic optimization, `make_optimizer_step` should be None.'
+                ' This option is enabled only with `Trainer(automatic_optimization=False)`'
+                ' Currently, all optimizers follow the `Trainer(accumulate_grad_batches=x)` logic')
+        self._accumulate_grad_batches = accumulate_grad_batches
 
     def _on_trainer_init(self, trainer):
         self._trainer = proxy(trainer)
+        self._automatic_optimization = trainer.train_loop.automatic_optimization
+        for opt_idx, opt in enumerate(trainer.optimizers):
+            if opt == self._optimizer:
+                self._optimizer_idx = opt_idx
+                break
+
+    @classmethod
+    def to_lightning_optimizer(cls, optimizer, trainer):
+        optimizer = cls(optimizer)
+        optimizer._on_trainer_init(trainer)
+        return optimizer
 
     def _accumulated_batches_reached(self):
-        if self._use_accumulate_grad_batches_from_trainer:
-            accumulate_grad_batches = self._trainer.accumulate_grad_batches
-        else:
-            accumulate_grad_batches = self._accumulate_grad_batches
-        return (self._trainer.batch_idx + 1) % accumulate_grad_batches == 0
+        if self.accumulate_grad_batches is None:
+            return self._trainer.train_loop._accumulated_batches_reached()
+        return (self._trainer.batch_idx + 1) % self.accumulate_grad_batches == 0
```

Comment:
isn't `_accumulate_grad_batches` always None here? I can't it being set to a value anywhere.

Reply:
Yes, _accumulate_grad_batches is None.
People will be able to set an accumulated value if they want.
```
optimizer = LightningOptimizer(optimizer, 3)

or 

optimizer.accumulate_grad_batches = 3
```

Or could even have random `.step`.
```
optimizer.step(make_optimzer_step=np.random.randint(100) > 90)
```